### PR TITLE
Src/tagger

### PR DESCRIPTION
**Pull request by celinethi**
add tagger.py 
when mode= test, evaluate trained model on new text and returns predicted labels and predicted scores for each labels for each words in text

### CODE REVIEW

File: src/dataset.py [hunk 1]
```diff
@@ -32,48 +34,76 @@ def __init__(
...
+            self.input_ids, self.attention_masks = (
+                torch.tensor(self.input_ids),
+                torch.tensor(self.attention_masks),
+            )
 
-        self.data = TensorDataset(self.input_ids, self.attention_masks, self.tags)
+            self.data = TensorDataset(self.input_ids, self.attention_masks)
 
-        self.len = len(self.labels)  # to check
+            self.len = len([[s for s in sent] for sent in getter.sentences])  # to check
```
Comment (jblaval) [inline src/dataset.py:106]: In the "train" part, `len(self.labels)` was used because `self.labels` was defined.
In this part, it seems to be simpler to simply write `self.len = len(getter.sentences)`.

File: src/tagger.py [hunk 2]
```diff
@@ -1,5 +1,112 @@
...
+
+import torch
+from torch import nn
+from transformers import AutoModelForTokenClassification, AutoConfig, AutoTokenizer
+import torch.nn.functional as F
+
+from src.models.bert_model_bis import BertForTokenClassificationModified, BertForTokenClassificationCRF
+
+
+# test git
```
Comment (jblaval) [inline src/tagger.py:15]: Why this ?

File: src/tagger.py [hunk 3]
```diff
@@ -1,5 +1,112 @@
...
+class Tagger:
+    def __init__(
+            self,
+            dataset,
+            batch_size=100,
+            path_previous_model=None,
+            saving_dir="data/results/",
```
Comment (jblaval) [inline src/tagger.py:22]: Tagger doesn't use `saving_dir`

File: src/tagger.py [hunk 4]
```diff
@@ -1,5 +1,112 @@
...
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+        self.batch_size = batch_size
+
+        self.path_previous_model = path_previous_model
+
+        self.dataset = dataset
+
+        self.saving_dir = Path(saving_dir)
```
Comment (jblaval) [inline src/tagger.py:36]: Tagger doesn't use `saving_dir`

File: src/dataset.py [hunk 5]
```diff
@@ -18,6 +19,7 @@ class NerDataset(Dataset):
     def __init__(
             self,
             data_path,
+            mode,
```
Comment (LaRiffle) [inline src/dataset.py:22]: Don't forget to add it in the docstring :)

File: src/dataset.py [hunk 6]
```diff
@@ -32,48 +34,76 @@ def __init__(
...
+                padding="post",
+            )
+
+            self.tags = pad_sequences(
+                [[self.tag2idx.get(l) for l in lab] for lab in self.labels],
+                maxlen=max_len,
+                value=self.tag2idx["O"],
+                padding="post",
+                dtype="long",
+                truncating="post",
```
Comment (LaRiffle) [inline src/dataset.py:66]: You can order the kwargs in the same order to improve readability

File: src/dataset.py [hunk 7]
```diff
@@ -32,48 +34,76 @@ def __init__(
...
+            self.input_ids = pad_sequences(
+                [tokenizer.convert_tokens_to_ids(txt) for txt in tokenized_texts],
+                maxlen=max_len,
+                dtype="long",
+                truncating="post",
+                padding="post",
+            )
 
-        self.attention_masks = [[float(i > 0) for i in ii] for ii in self.input_ids]
+            self.attention_masks = [[float(i > 0) for i in ii] for ii in self.input_ids]
```
Comment (LaRiffle) [inline src/dataset.py:97]: Because we convert at the end to pytorch you can leverage pytorch functionalities to have faster execution:
```python
input_ids= [[-1, 2], [-5, 3]]
# option 1
attention_masks = [[float(i > 0) for i in ii] for ii in input_ids]
torch.tensor(attention_masks)
# option 2
(torch.tensor(input_ids) > 0).float()
```

File: src/dataset.py [hunk 8]
```diff
@@ -32,48 +34,76 @@ def __init__(
...
+
+            self.input_ids, self.tags, self.attention_masks = (
+                torch.tensor(self.input_ids),
+                torch.tensor(self.tags),
+                torch.tensor(self.attention_masks),
+            )
+
+            self.data = TensorDataset(self.input_ids, self.attention_masks, self.tags)
+
+            self.len = len(self.labels)  # to check
```
Comment (LaRiffle) [inline src/dataset.py:79]: Is this still a "to check" or it's good?

File: src/dataset.py [hunk 9]
```diff
@@ -224,3 +254,49 @@ def get_next(self):
             return s
         except:
             return None
+
+
+class SentenceGetter_test(object):
```
Comment (LaRiffle) [inline src/dataset.py:259]: How different is it from the normal `SentenceGetter` ? If there is a difference, maybe we can use the same class but add a kwarg to specify train / test mode ?

File: src/tagger.py [hunk 10]
```diff
@@ -1,5 +1,108 @@
...
+
+
+class Tagger:
+    def __init__(
+            self,
+            dataset,
+            batch_size=100,
+            path_previous_model=None,
+    ):
+        random.seed(42)
```
Comment (LaRiffle) [inline src/tagger.py:21]: Don't forget to add a docstring explaining the behaviour of this class

File: src/tagger.py [hunk 11]
```diff
@@ -1,5 +1,108 @@
...
+                outputs = self.model(input_ids, token_type_ids=None, attention_mask=mask)
+                scores = F.softmax(outputs[0], dim=2)
+                for row in range(len(data)):
+                    df_predicted_label, df_predicted_score = self.get_words_back(input_ids[row], scores[row])
+                    prediction_label = pd.concat([prediction_label, df_predicted_label], axis=0)
+                    prediction_score = pd.concat([prediction_score, df_predicted_score], axis=0, sort=False)
+
+        return prediction_label, prediction_score
+
+    def get_words_back(self, input, score):
```
Comment (LaRiffle) [inline src/tagger.py:81]: Don't forget to add a docstring as well :)

File: src/tagger.py [hunk 12]
```diff
@@ -1,5 +1,108 @@
...
+        label_indices = np.argmax(score.to('cpu').numpy(), axis=1)
+        score = score.to('cpu').numpy()
+        # join bpe split tokens
+        tokenizer = AutoTokenizer.from_pretrained(self.pretrained_model, do_lower_case=True)
+        tokens = tokenizer.convert_ids_to_tokens(input.to('cpu').numpy())
+
+        sep = tokens[0][0]
+        new_tokens, new_labels, new_score = [], [], []
+        for token, label_idx, score in zip(tokens, label_indices, [score[i] for i in range(len(score))]):
+            if not (token.startswith(sep)):
```
Comment (LaRiffle) [inline src/tagger.py:93]: Not sure I understand this part, could you a comment?

File: src/tagger.py [hunk 13]
```diff
@@ -1,5 +1,108 @@
...
+                columns=['words'] + [f'p_{self.config.id2label[id]}' for id in range(self.num_labels)])
+            for data in self.test_loader:
+                input_ids, mask = data
+                input_ids = input_ids.to(self.device)
+                mask = mask.to(self.device)
+                outputs = self.model(input_ids, token_type_ids=None, attention_mask=mask)
+                scores = F.softmax(outputs[0], dim=2)
+                for row in range(len(data)):
+                    df_predicted_label, df_predicted_score = self.get_words_back(input_ids[row], scores[row])
+                    prediction_label = pd.concat([prediction_label, df_predicted_label], axis=0)
```
Comment (LaRiffle) [inline src/tagger.py:76]: I would suggest here to stay in the pytorch world: using pandas forces you to go back to cpu and then to numpy and then to pandas, which is a lot of conversions. Could you keep the conversion just for the very end of this function, and the return,  (if you want to return a df which is more readable) ?

Also, I guess all this processing can be done out of the `for data in self.test_loader:` , especially the concat part: you could do a huge torch.stack on a list of pytorch tensors. For the `self.get_words_back` as well probably

File: src/dataset.py [hunk 14]
```diff
@@ -28,26 +28,44 @@ def __init__(
...
+            for sent in getter.sentences:
+                words = [s[0] for s in sent]
+                labels = [s[1] for s in sent]
+                tokens = []
+                label_ids = []
+                for word, label in zip(words, labels):
+                    word_tokens = tokenizer.tokenize(word)
+                    tokens.extend(word_tokens)
+                    # Use the real label id for the first token of the word, and padding ids for the remaining tokens
+                    #label_ids.extend([self.tag2idx.get(label)] + [self.pad_token_label_id] * (len(word_tokens) - 1))
```
Comment (LaRiffle) [inline src/dataset.py:64]: you can remove the first option then :) 
```suggestion
```

File: src/dataset.py [hunk 15]
```diff
@@ -215,12 +228,13 @@ class SentenceGetter(object):
     Data extractor from .csv file
     """
 
-    def __init__(self, data_path, encoding="latin1"):
+    def __init__(self, mode, data_path, encoding="latin1"):
```
Comment (LaRiffle) [inline src/dataset.py:231]: So should we default to latin1 or utf-8?

File: src/dataset.py [hunk 16]
```diff
@@ -58,28 +76,26 @@ def __init__(
             )
 
             self.tags = pad_sequences(
-                [[self.tag2idx.get(l) for l in lab] for lab in self.labels],
+                tag_ids,
                 maxlen=max_len,
                 value=self.tag2idx["O"],
-                padding="post",
```
[the remark this answers is not in the document]
Comment (celinethi) [reply src/dataset.py]: yes default is "pre", but "post" is preferred for BERT

File: src/dataset.py (hunk 15 above)
Comment (celinethi) [reply src/dataset.py:231]: It depends on the input text, maybe we should add it as a new pypa.py argument ?
